scalpel.1
.TH SCALPEL "1" "v1.60 - December 2006" "A Fast Header/Footer-based File Carver" "A Fast Header/Footer-based File Carver" 

.SH NAME
scalpel \- Recover files using a header/footer database

.SH SYNOPSIS
.B scalpel
[\fB-b\fR]
[\fB-c\fR <file>]
[\fB-d\fR]
[\fB-h\fR]
[\fB-i\fR <file>]
[\fB-m\fR <blocksize>]
[\fB-n\fR]
[\fB-o\fR <dir>] 
[\fB-O\fR]
[\fB-p\fR]
[\fB-r\fR]
[\fB-s\fR <num>]
[\fB-t\fR]
[\fB-u\fR]
[\fB-V\fR]
[\fB-v\fR]
[\fIFILES\fR]...

.SH DESCRIPTION
.PP
Recover files from a disk image or raw block device based on headers 
and footers specified by the user.

.TP
\fB\-b\fR
Carve files even if defined footers aren't discovered within
maximum carve size for file type [foremost 0.69 compat mode]

.TP
\fB-c\fR \fIfile\fR
Chooses which configuration file to use. If this option is omitted,
then "scalpel.conf" in the current directory is used. The format for
the configuration file is described in the default configuration
file "scalpel.conf".  See the \fICONFIGURATION FILE\fR
section below for more information.

.TP
\fB\-d\fR
Generate header/footer database; will bypass certain optimizations
and discover all footers, so performance suffers.  Doesn't affect
the set of files carved.  **EXPERIMENTAL**

.TP
\fB\-m\fR
Generate/update carve coverage blockmap file.  The first 32bit
unsigned int in the file identifies the block size. Thereafter
each 32bit unsigned int entry in the blockmap file corresponds
to one block in the image file.  Each entry counts how many
carved files contain this block. Requires more memory and
disk.  **EXPERIMENTAL**

.TP
\fB\-h\fR
Show a help screen and exit.

.TP
\fB\-i\fR \fIfile\fR
\fIfile\fR is used as a list of input files to examine. Each
line in the specified file should contain a single filename.

.TP
\fB-o\fR \fIdirectory\fR
Recovered files are written to the directory
\fIdirectory\fR.   Scalpel requires that this directory
be either empty or not exist.  The directory will be created
if necessary.

.TP
\fB\-O\fR
Don't organize carved files by type. Default is to organize carved files
into subdirectories to make previewing of large numbers of carved
files easier.

.TP
\fB\-p\fR
Perform image file preview; audit log indicates which files
would have been carved, but no files are actually carved.

.TP
\fB\-q\fR \fIclustersize\fR
Carve only when header is cluster-aligned.

.TP
\fB\-r\fR
Find only first of overlapping headers/footers [foremost 0.69 compat mode]

.TP
\fB-s\fR \fInumber\fR
Skips \fInumber\fR bytes in each input file before beginning the search
for file headers and footers.

.TP
\fB\-t\fR
Set directory for coverage blockmap.  **EXPERIMENTAL**

.TP
\fB\-u\fR
Use carve coverage blockmap when carving.  Carve only sections
of the image whose entries in the blockmap are 0.  These areas
are treated as contiguous regions.  **EXPERIMENTAL**

.TP
\fB\-V\fR
Show copyright information and exit.

.TP
\fB\-v\fR
Enables verbose mode. This causes copious amounts of debugging information
to be output.

.PP

.SH CONFIGURATION FILE
The configuration file is used to control the types of files Scalpel
will attempt to carve.  A sample configuration file, "scalpel.conf", is included with
this distribution. For each file type, the configuration file describes
the file's extension, whether the header and footer are case sensitive,
the maximum file size, and the header and footer for the file. The footer
field is optional, but the header, size, case sensitivity, and extension fields
are required. 

Important note: The default configuration file has all supported file
patterns commented out--you must edit this before before running
Scalpel.

Any line in the configuration file that begins with a pound sign is 
considered a comment and ignored. 

Headers and footers are decoded before use. To specify a value in
hexadecimal use \\x[0-f][0-f], and for octal use \\[1-9][1-9][1-9].  Spaces
can be represented by \\s. Example: "\\x4F\\123\\I\\sCCI" decodes to "OSI CCI".

To match any single character (aka a wildcard) use a '?'. If you need to
search for the '?' character, you will need to change the 'wildcard' line
*and* every occurrence of the old wildcard character in the configuration
file, including those appearing in hex and octal values. '?' is equal to \\x3f and
\\063.

.SH AUTHORS
Written by Golden G. Richard III.  The first version of Scalpel was based
on foremost 0.69, which was written by Special Agent Kris Kendall and 
Special Agent Jesse Kornblum of the United States Air Force 
Office of Special Investigations.

.SH BUGS AND LIMITATIONS
It is currently not possible to carve physical block devices directly using
the Windows version of Scalpel.  This is a limitation that will be removed in 
a future release of Scalpel.

.SH "REPORTING BUGS"
When submitting a bug report, please include a description
of the problem, how you found it, and your contact information.
.PP
Send bug reports to:
.br
goldenrichard@gmail.com
.PP
.SH COPYRIGHT
.PP
This is free software.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

.SH "SEE ALSO"
More information on Scalpel appears in the README file, distributed
with the Scalpel source code.
.PP


changelog
1.4:	First public beta release.
1.5:	Added "-b" option for complete carving compability with 
	foremost 0.69.
1.51:	Fixed a problem with Scalpel not carving all files near the end of 
	some large images. The locations of the files was detected, but
	an optimization to minimize reads on pass 2 caused the files not to
	be written (and not to be noted in the Scalpel log).
	Also, more work to bring the Win32 version into compliance.  :(
	Fixed problem with Win32 version reporting 
	"Couldn't measure size of image file" for large images--required 
	ftello64()/fseeko64() under mingw.  40+GB images have now been
	carved successfully under Win32.
	Fixed format of offset in log file for large offsets under Win32;
	the printf() implementation in mingw under Windows doesn't support
	"%llu" for unsigned long long integers--needs "%I64u".  Blech.
1.52	Foremost 0.69 didn't discover overlapping headers/footers.  If you want
	this behavior, use "-r" on the command line.  Scalpel's default 
	behavior is now to discover all headers/footers, even if they overlap.
	Different Quicktime needles (from the 'file' command's magic file) 
	are now used in the configuration file, because the default Foremost 
	needles didn't seem to match valid Quicktime files.  The needles 
	are still weak, in that they cause a bunch of false positives.  
	The configuration file now uses more reasonable (in my opinion) 
	maximum carve sizes.  By default, all file types are still 
	commented out.
1.53	Verbose mode implemented.  "-v" outputs copious (too much?) information
	about what Scalpel is doing as files are carved. 
	Mac OS X is now supported, including the use of raw block devices
	(e.g., /dev/disk0) as carve targets.   Use "make bsd" for an 
	OS X compilation.  
1.54	Maximum size for carved files is now 18 exabytes. Minor changes to
	Makefile. Fixed minor formatting problem in audit log.  Scalpel
	man page created.  Eliminated duplicate error messages when
	targets could not be opened.  Thanks to John Vigo for additional
	MAC OS X testing on this release.
1.60	Changes in this release include:

	o Some of the headers and footers in the default
	'scalpel.conf' file have been improved.  In particular, the
	GIF89 footer was made less restrictive; GIF89 files were
	encountered in the wild for which the legacy Foremost footer
	was too restrictive.

	o REVERSE carving semantics now fixed as "helpers.c"
	functions continue to evolve from their Foremost roots; this
	is mostly useful for PDF carving and when some carved files
	appear to be truncated--e.g., many times GIF files will
	contain the footer string internally and with FORWARD carving,
	will be truncated.  REVERSE carving can help, but at the
	expense of occasionally carving overly large files.
	
	o Removed obsolete option "-f".

	o Maximum number of files that can be carved per file type is
	now 18,446,744,073,709,551,616.  This is also the maximum
	number of files of all combined types that can be carved in
	a single run.  This is unlikely to be a problematic upper
	bound in the near term.

	o The "-q" command line option is now available to force
	block-aligned carving; only files with headers that begin on
	block boundaries are carved.  The block size for this mode is
	user-specified.

	o The "-r" option was not accepted in recent releases due to
	an omission in Scalpel's command line argument parsing.  This
	option isn't generally used except for compatibility testing
	with Foremost 0.69 (it actually forces Scalpel to mimic some
	buggy Foremost behavior) and should be considered deprecated.
	In a future release of Scalpel, it will be removed.

	o Scalpel now organizes carved files into subdirectories by
	type (and for each type, additional subdirectories are created
	with a maximum of MAX_FILES_PER_SUBDIRECTORY [default: 5000]
	files each).  This will allow you to more comfortably explore
	the set of carved files using the Windows or Linux file
	explorers without having them choke on large subdirectories.
	The files are organized within the Scalpel output directory
	into subdirectories with the following naming convention:

	suffix-rule#-counter  
	suffix-rule#-counter  
	...
	suffix-rule#-counter  

	where 'suffix' matches the suffix specified in the Scalpel
	configuration file (e.g, "jpg"), 'rule#' is the index of the
	corresponding rule in the configuration file (with the first
	uncommented rule being 0), and 'counter' being incremented after
	every MAX_FILES_PER_SUBDIRECTORY files corresponding to a rule are
	carved.
	
	You can turn off this default organization by using the "-O"
	command line flag (then Scalpel acts as in previous releases
	and dumps all carved files into a single output directory).  

	o The "-p" performs a "preview" carve, creating an audit file
	indicating which files would be carved, but without performing
	carving operations.  This substantially increases execution speed
	and also supports "in-line" carving, when used in conjunction with
	our custom FUSE filesystem.

	o A memory corruption error when Scalpel had large numbers of
	files open (typically, when a large number of file types were
	specified in the configuration file) was corrected.  Thanks to

scalpel.c
// Scalpel Copyright (C) 2005-6 by Golden G. Richard III.
// Written by Golden G. Richard III.
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as
// published by the Free Software Foundation; either version 2 of the
// License, or (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
// 02110-1301, USA.

// Scalpel is a complete rewrite of the foremost 0.69 file carver to
// increase speed and support execution on machines with minimal
// resources (e.g., < 256MB RAM).
//
// Thanks to Kris Kendall, Jesse Kornblum, et al for their work on
// foremost.  Additional thanks to Vassil Roussev and Vico Marziale, of the
// Department of Computer Science at the University of New Orleans, for 
// ongoing support of Scalpel.
//
// Scalpel currently uses the SAME configuration file format as
// foremost 0.69.  This may change in the future, because of some
// planned new features.
//

#include "scalpel.h"

// GLOBALS
int signal_caught; 
char wildcard;
int ttywidth;
char *__progname;

void usage() {
  
  printf("Carves files from a disk image based on file headers and footers.\n");
  printf("\nUsage: scalpel [-b] [-c <config file>] [-d] [-h|V] [-i <file>]\n");
  printf("                 [-m blocksize] [-n] [-o <outputdir>] [-O num] [-q clustersize]\n");
  printf("                 [-r] [-s num] [-t <blockmap file>] [-u] [-v]\n"); 
  printf("                 <imgfile> [<imgfile>] ...\n\n");
  printf("-b  Carve files even if defined footers aren't discovered within\n");
  printf("    maximum carve size for file type [foremost 0.69 compat mode].\n");
  printf("-c  Choose configuration file.\n");
  printf("-d  Generate header/footer database; will bypass certain optimizations\n");
  printf("    and discover all footers, so performance suffers.  Doesn't affect\n");
  printf("    the set of files carved.  **EXPERIMENTAL**\n");
  printf("-h  Print this help message and exit.\n");
  printf("-i  Read names of disk images from specified file.\n");
  printf("-m  Generate/update carve coverage blockmap file.  The first 32bit\n");
  printf("    unsigned int in the file identifies the block size. Thereafter\n");
  printf("    each 32bit unsigned int entry in the blockmap file corresponds\n");
  printf("    to one block in the image file.  Each entry counts how many\n");
  printf("    carved files contain this block. Requires more memory and\n");
  printf("    disk.  **EXPERIMENTAL**\n");
  printf("-n  Don't add extensions to extracted files.\n");
  printf("-o  Set output directory for carved files.\n");
  printf("-O  Don't organize carved files by type. Default is to organize carved files\n");
  printf("    into subdirectories.\n");
  printf("-p  Perform image file preview; audit log indicates which files\n");
  printf("    would have been carved, but no files are actually carved.\n");
  printf("-q  Carve only when header is cluster-aligned.\n"); 
  printf("-r  Find only first of overlapping headers/footers [foremost 0.69 compat mode].\n");
  printf("-s  Skip n bytes in each disk image before carving.\n");
  printf("-t  Set directory for coverage blockmap.  **EXPERIMENTAL**\n");
  printf("-u  Use carve coverage blockmap when carving.  Carve only sections\n");
  printf("    of the image whose entries in the blockmap are 0.  These areas\n");
  printf("    are treated as contiguous regions.  **EXPERIMENTAL**\n");
  printf("-V  Print copyright information and exit.\n");
  printf("-v  Verbose mode.\n");
}


// signal handler, sets global variable 'signal_caught' which is
// checked periodically during carve operations.  Allows clean
// shutdown.
void catch_alarm (int signum) {
  signal_caught = signum;
  signal(signum,catch_alarm);

#ifdef __DEBUG
  fprintf(stderr,"\nCaught signal: %s.\n",(char*) strsignal(signum));
#endif

  fprintf (stderr, "\nKill signal detected. Cleaning up...\n");
}


int extractSearchSpecData(struct SearchSpecLine *s,char **tokenarray) {

  // process one line from config file:
  //     token[0] = suffix
  //     token[1] = case sensitive?
  //     token[2] = maximum carve size
  //     token[3] = begintag
  //     token[4] = endtag
  //     token[5] = search type (optional)

  s->suffix = malloc(MAX_SUFFIX_LENGTH*sizeof(char));
  s->begin  = malloc(MAX_STRING_LENGTH*sizeof(char));
  s->end    = malloc(MAX_STRING_LENGTH*sizeof(char));    
  
  if (!strncasecmp(tokenarray[0],
                   SCALPEL_NOEXTENSION_SUFFIX,
                   strlen(SCALPEL_NOEXTENSION_SUFFIX))) {
    s->suffix[0] = SCALPEL_NOEXTENSION;
    s->suffix[1] = 0;
  }
  else {
    memcpy(s->suffix,tokenarray[0],MAX_SUFFIX_LENGTH);
  }
  
  // case sensitivity check
  s->casesensitive = (!strncasecmp(tokenarray[1],"y",1) || 
		      !strncasecmp(tokenarray[1],"yes",3));
  
  //#ifdef __WIN32
  //    s->length = _atoi64(tokenarray[2]);
  //#else
  //  s->length = atoull(tokenarray[2]);
  //#endif

  s->length = strtoull(tokenarray[2], 0, 10);

  // determine search type for this needle
  s->searchtype = SEARCHTYPE_FORWARD;
  if (!strncasecmp(tokenarray[5],"REVERSE",strlen("REVERSE"))) {
    s->searchtype = SEARCHTYPE_REVERSE;
  }
  else if (!strncasecmp(tokenarray[5],"NEXT",strlen("NEXT"))) {
    s->searchtype = SEARCHTYPE_FORWARD_NEXT;
  }
  // FORWARD is the default, but OK if the user defines it explicitly
  else if (!strncasecmp(tokenarray[5],"FORWARD",strlen("FORWARD"))) {
    s->searchtype = SEARCHTYPE_FORWARD;
  }

  s->beginlength = translate(tokenarray[3]);
  memcpy(s->begin,tokenarray[3],s->beginlength);
  s->endlength = translate(tokenarray[4]);
  memcpy(s->end,tokenarray[4],s->endlength);  
 
  init_bm_table(s->begin,s->begin_bm_table,s->beginlength, 
		s->casesensitive);
  init_bm_table(s->end,s->end_bm_table,s->endlength,
		s->casesensitive);
  return SCALPEL_OK;
}


int processSearchSpecLine(struct scalpelState *state, char *buffer, 
			  int lineNumber) {
  
  char *buf = buffer;
  char *token;
  char **tokenarray = (char **) malloc(6*sizeof(char[MAX_STRING_LENGTH+1]));
  int i = 0, err = 0, len = strlen(buffer);

  // murder CTRL-M (0x0d) characters that terminate a line
  if (len >= 2 && buffer[len-2] == 0x0d && buffer[len-1] == 0x0a) {  
    buffer[len-2] = buffer[len-1];
    buffer[len-1] = buffer[len];
  }

  buf = (char*) skipWhiteSpace(buf);
  token = strtok(buf," \t\n");

  // lines beginning with # are comments
  if(token == NULL || token[0] == '#'){  
    return SCALPEL_OK;
  }
    
  // allow wildcard to be changed
  if (!strncasecmp(token,"wildcard",9)) {
    if ((token = strtok(NULL," \t\n")) != NULL) {
      translate(token);
    } 
    else {
      fprintf (stdout,"Warning: Empty wildcard in configuration file line %d. Ignoring.\n",
	       lineNumber);
      return SCALPEL_OK;
    }

    if (strlen(token) > 1) {
      fprintf(stderr,"Warning: Wildcard can only be one character,"
	      " but you specified %d characters.\n"
	      "         Using the first character, \"%c\", as the wildcard.\n",
	      (int)strlen(token),token[0]);
    }

    wildcard = token[0];
    return SCALPEL_OK;
  }
    
  while (token && (i < NUM_SEARCH_SPEC_ELEMENTS)){
    tokenarray[i] = token;
    i++;
    token = strtok(NULL," \t\n");
  }
  
  switch(NUM_SEARCH_SPEC_ELEMENTS-i){
    case 2:
      tokenarray[NUM_SEARCH_SPEC_ELEMENTS-1] = "";
      tokenarray[NUM_SEARCH_SPEC_ELEMENTS-2] = "";
      break;
    case 1:
      tokenarray[NUM_SEARCH_SPEC_ELEMENTS-1] = "";
      break;
    case 0:
      break;
    default:
      fprintf(stderr, 
	      "\nERROR: In line %d of the configuration file, expected %d tokens,\n"
	      "       but instead found only %d.\n", 
	      lineNumber,NUM_SEARCH_SPEC_ELEMENTS,i);
      return SCALPEL_ERROR_NO_SEARCH_SPEC;
      break;
      
  }

  if ((err = extractSearchSpecData(&(state->SearchSpec[state->specLines]),
				   tokenarray))) {
    switch(err) {
    default:
      fprintf(stderr,
	      "\nERROR: Unknown error on line %d of the configuration file.\n"
	      ,lineNumber);
    }
  }
  state->specLines++; 
  return SCALPEL_OK;
}


// process configuration file
int readSearchSpecFile(struct scalpelState *state) {

  int lineNumber = 0, status;
  FILE *f;
  
  char *buffer = malloc((NUM_SEARCH_SPEC_ELEMENTS * MAX_STRING_LENGTH + 1) * sizeof(char));

  f = fopen(state->conffile,"r");  
  if (f == NULL) {
    fprintf (stderr,
	     "ERROR: Couldn't open configuration file: %s -- %s\n", 
	     state->conffile,strerror(errno));
    free(buffer);
    return SCALPEL_ERROR_FILE_OPEN;
  }

  while (fgets(buffer,NUM_SEARCH_SPEC_ELEMENTS * MAX_STRING_LENGTH, f)) {
    lineNumber++;

    if (state->specLines > MAX_FILE_TYPES) {
      fprintf(stderr,"Your conf file contains too many file types.\n");
      fprintf(stderr,"This version was compiled with MAX_FILE_TYPES == %d.\n",
              MAX_FILE_TYPES);
      fprintf(stderr,"Increase MAX_FILE_TYPES, recompile, and try again.\n");
      free(buffer);
      return SCALPEL_ERROR_TOO_MANY_TYPES;
    }

    if ((status = processSearchSpecLine(state,buffer,lineNumber)) != SCALPEL_OK) {
      free(buffer);
      return status;
    }
  }

  // add an empty object to the end of the list as a marker

  state->SearchSpec[state->specLines].suffix = NULL;
  state->SearchSpec[state->specLines].casesensitive = 0;
  state->SearchSpec[state->specLines].length = 0;
  state->SearchSpec[state->specLines].begin = NULL;
  state->SearchSpec[state->specLines].beginlength = 0;
  state->SearchSpec[state->specLines].end = NULL;
  state->SearchSpec[state->specLines].endlength = 0;

  // GGRIII: offsets field is uninitialized--it doesn't
  // matter, since we won't use this entry.
  
  fclose(f);
  free(buffer);
  return SCALPEL_OK;
}

// Register the signal-handler that will write to the audit file and
// close it if we catch a SIGINT or SIGTERM 
void registerSignalHandlers() {
  if (signal (SIGINT, catch_alarm) == SIG_IGN) {
      signal (SIGINT, SIG_IGN);
  }
  if (signal (SIGTERM, catch_alarm) == SIG_IGN) {
    signal (SIGTERM, SIG_IGN);
  }

#ifndef __WIN32
    // *****GGRIII:  is this problematic?
    // From foremost 0.69:
    /* Note: I haven't found a way to get notified of 
       console resize events in Win32.  Right now the statusbar
       will be too long or too short if the user decides to resize 
       their console window while foremost runs.. */

    signal(SIGWINCH, setttywidth);
#endif 
}

// initialize state variable and copy command line arguments
void initializeState(char **argv, struct scalpelState *state) {
  
  char** argvcopy = argv;
  int sss;
  int i;
  
  // Allocate memory for the state 
  state->imagefile        = (char*) malloc(MAX_STRING_LENGTH * sizeof(char));
  state->inputFileList    = (char*) malloc(MAX_STRING_LENGTH * sizeof(char));
  state->conffile         = (char*) malloc(MAX_STRING_LENGTH * sizeof(char));
  state->outputdirectory  = (char*) malloc(MAX_STRING_LENGTH * sizeof(char));
  state->invocation       = (char*) malloc(MAX_STRING_LENGTH * sizeof(char));

  // GGRIII: memory allocation made more sane, because we're storing
  // more information in Scalpel than foremost had to, for each file
  // type.
  sss = (MAX_FILE_TYPES+1)*sizeof(struct SearchSpecLine);
  state->SearchSpec = (struct SearchSpecLine*) malloc(sss);
  state->specLines = 0;

  // GGRIII: initialize header/footer offset data, carved file count,
  // et al.  The header/footer database is re-initialized in "dig.c"
  // after each image file is processed (numfilestocarve and
  // organizeDirNum are not). Storage for the header/footer offsets
  // will be reallocated as needed.

  for (i=0; i < MAX_FILE_TYPES; i++) {
    state->SearchSpec[i].offsets.headers = 0;
    state->SearchSpec[i].offsets.footers = 0;
    state->SearchSpec[i].offsets.numheaders = 0;
    state->SearchSpec[i].offsets.numfooters = 0;
    state->SearchSpec[i].offsets.headerstorage = 0;
    state->SearchSpec[i].offsets.footerstorage = 0;
    state->SearchSpec[i].numfilestocarve = 0;
    state->SearchSpec[i].organizeDirNum = 0;
  }

  state->fileswritten = 0;
  state->skip = 0;
  state->organizeMaxFilesPerSub = MAX_FILES_PER_SUBDIRECTORY;
  state->modeVerbose = FALSE;
  state->modeNoSuffix = FALSE;
  state->useInputFileList = FALSE;
  state->carveWithMissingFooters = FALSE;
  state->noSearchOverlap = FALSE;
  state->generateHeaderFooterDatabase = FALSE;
  state->updateCoverageBlockmap = FALSE;
  state->useCoverageBlockmap = FALSE;
  state->blockAlignedOnly = FALSE; 
  state->organizeSubdirectories = TRUE;
  state->previewMode = FALSE;
  state->ignoreEmbedded = FALSE;
  state->auditFile = NULL;

  // default values for output directory, config file, wildcard character,
  // coverage blockmap directory
  strncpy(state->outputdirectory,SCALPEL_DEFAULT_OUTPUT_DIR,
	  MAX_STRING_LENGTH);
  strncpy(state->conffile,SCALPEL_DEFAULT_CONFIG_FILE,
	  MAX_STRING_LENGTH);
  state->coveragedirectory = state->outputdirectory;
  wildcard = SCALPEL_DEFAULT_WILDCARD;
  signal_caught = 0;
  state->invocation[0] = 0;
  
  // copy the invocation string into the state
  do {
    strncat(state->invocation,  
	    *argvcopy, 
	    MAX_STRING_LENGTH-strlen(state->invocation));
    strncat(state->invocation,
	    " ",
	    MAX_STRING_LENGTH-strlen(state->invocation));
    ++argvcopy;  
  } while (*argvcopy);  

  registerSignalHandlers();
}

// parse command line arguments
void processCommandLineArgs(int argc, char **argv, 
			    struct scalpelState *state) {
  int i;

  while ((i = getopt(argc, argv, "bhvVundpq:rt:c:o:s:i:m:O")) != -1) {
    switch (i) {
	
    case 'V':
      fprintf (stdout,SCALPEL_COPYRIGHT_STRING);
      exit(1);
      
    case 'h':
      usage();
      exit(1);
      
    case 's':
      state->skip = strtoull(optarg,NULL,10);
      fprintf (stdout,"Skipping the first %lld bytes of each image file.\n",
	       state->skip);
      break;
      
    case 'c':
      strncpy(state->conffile,optarg,MAX_STRING_LENGTH);
      break;

    case 'd':
      state->generateHeaderFooterDatabase = TRUE;
      break;

      // -e support is currently a work-in-progress and will be enabled in a future release
      //    case 'e':
      //      state->ignoreEmbedded=TRUE;
      //      break;

    case 'm':
      state->updateCoverageBlockmap = TRUE;
      state->coverageblocksize=strtoul(optarg,NULL,10);
      if (state->coverageblocksize <= 0) {
	fprintf(stderr, 
		"\nERROR: Invalid blocksize for -m command line option.\n");
	exit(1);
      }
      break;
      
    case 'o':
      strncpy(state->outputdirectory,optarg,MAX_STRING_LENGTH);
      break;

    case 't':
      state->coveragedirectory = (char *) malloc(MAX_STRING_LENGTH * sizeof(char));
      strncpy(state->coveragedirectory,optarg,MAX_STRING_LENGTH);
      break;

    case 'O':
      state->organizeSubdirectories = FALSE;
      break;

    case 'p':
      state->previewMode = TRUE;
      break;

    case 'b':
      state->carveWithMissingFooters = TRUE;
      break;
      
    case 'i':
      state->useInputFileList = TRUE;
      state->inputFileList = optarg;  
      break;	  
      
    case 'n':
      state->modeNoSuffix = TRUE;
      fprintf (stdout,"Extracting files without filename extensions.\n");
      break;

    case 'q':
      state->blockAlignedOnly = TRUE;
      state->alignedblocksize = strtoul(optarg,NULL,10);
      if (state->alignedblocksize <= 0) {
	fprintf(stderr, 
		"\nERROR: Invalid blocksize for -q command line option.\n");
	exit(1);
      }
      break;

    case 'r':
      state->noSearchOverlap = TRUE;
      break;

    case 'u':
      state->useCoverageBlockmap = TRUE;
      break;

    case 'v':
      state->modeVerbose = TRUE;
      break;

    default:
      exit(1);
    }
  }
}

// full pathnames for all files used
void convertFileNames(struct scalpelState *state) {

  char fn[MAX_STRING_LENGTH];  

  realpath(state->outputdirectory,fn);
  strncpy(state->outputdirectory,fn,MAX_STRING_LENGTH);

  realpath(state->conffile,fn);
  strncpy(state->conffile,fn,MAX_STRING_LENGTH);
}


// GGRIII: for each file, build header/footer offset database first,
// then carve files based on this database.  Need to clear the
// header/footer offset database after processing of each file.

void digAllFiles(int argc, char **argv, struct scalpelState *state) {

  int i = 0, j = 0; 
  FILE* listoffiles = NULL;

  if (state->useInputFileList) {
    fprintf(stdout, "Batch mode: reading list of images from %s.",
	    state->inputFileList);
    listoffiles = fopen(state->inputFileList,"r");
    if (listoffiles == NULL) {
      fprintf(stderr, "Couldn't open file: %s -- %s\n", 
	      (*(state->inputFileList) == '\0')?"<blank>":state->inputFileList,
	      strerror(errno));
      closeFile(state->auditFile);
      exit(-1);
    }
    j = 0;
    do {
      j++;
      
      if (fgets(state->imagefile,MAX_STRING_LENGTH,listoffiles) == NULL) {
	
	fprintf(stderr,
		"Error reading line %d of %s. Skipping line.\n", 
		j,state->inputFileList);
	continue;
      }
      if(state->imagefile[strlen(state->imagefile)-1] == '\n'){
	state->imagefile[strlen(state->imagefile)-1] = '\x00';
      }

      // GGRIII: this function now *only* builds the header/footer
      // database.  Carving is handled afterward, in carveImageFile().
      
      if ((i = digImageFile(state))) {
	handleError(state,i);
      }
      else {
	// GGRIII: "digging" is now complete and header/footer database
	// has been built.  The function carveImageFile() performs
	// extraction of files based on this database.
      
	if ((i = carveImageFile(state))) {
	  handleError(state,i);
	}		
      }
    } 
    while (!feof(listoffiles));
    closeFile(listoffiles);
  }
  else {
    do {
      state->imagefile = *argv;
      
      // GGRIII: this function now *only* builds the header/footer
      // database.  Carving is handled afterward, in carveImageFile().

      if ((i = digImageFile(state))) {
	handleError(state,i);
      }		
      else {
	// GGRIII: "digging" is now complete and header/footer database
	// has been built.  The function carveImageFile() performs extraction 
	// of files based on this database.
	
	if ((i = carveImageFile(state))) {
	  handleError(state,i);
	}		
      }
      ++argv;  
    } while (*argv);
  }
}


int main(int argc, char **argv) {

  time_t starttime = time(0);
  struct scalpelState state;

  if (ldiv(SIZE_OF_BUFFER,SCALPEL_BLOCK_SIZE).rem != 0) {
    fprintf (stderr, SCALPEL_SIZEOFBUFFER_PANIC_STRING);
    exit (-1);
  }
  
#ifndef __GLIBC__
  setProgramName(argv[0]);
#endif

  fprintf (stdout,SCALPEL_BANNER_STRING);

  initializeState(argv,&state);

  processCommandLineArgs(argc,argv,&state);

  convertFileNames(&state);

  if (state.modeVerbose) {
    fprintf (stdout,"Output directory: \"%s\"\n", state.outputdirectory);
    fprintf (stdout,"Configuration file: \"%s\"\n", state.conffile);
    fprintf (stdout,"Coverage maps directory: \"%s\"\n", state.coveragedirectory);
  }

  // read configuration file
  if (readSearchSpecFile(&state)) {
    // error in configuration file, msg has already been output
    exit(-1);
  }

  setttywidth();

  argv += optind;
  if (*argv != NULL || state.useInputFileList) {
    // prepare audit file and make sure output directory is empty.
    if(openAuditFile(&state)){
      fprintf(stderr, "Aborting.\n\n");
      exit(-1);
    }
    digAllFiles(argc,argv,&state);
    closeFile(state.auditFile);
  } else {      
    usage();
    fprintf(stdout,"\nERROR: No image files specified.\n\n");
  }

#ifdef __WIN32
  fprintf (stdout,"\nScalpel is done, files carved = %I64u, elapsed = %ld seconds.\n", 
	   state.fileswritten, 
	   (int)time(0) - starttime);
#else
  fprintf (stdout,"\nScalpel is done, files carved = %llu, elapsed = %ld seconds.\n", 
	   state.fileswritten, 
	   (int)time(0) - starttime);
#endif

  return 0;
}
